Add FM30 TX hardware support

CapnBry opened the pull request:
This implements the transmitter portion of #381, support for FM30 TX hardware.

## Things to consider
* The bootloader does not work other than launching into the transmitter application. I've spent 20 hours on it and can't figure out what is wrong. OpenTX reports NoSync, although we do receive some of the STK sync. I receive more if I change [this highly suspect line](https://github.com/ExpressLRS/ExpressLRS/blob/develop/src/bootloader/src/Src/uart.c#L554) in the bootloader to ```usart_pin_config(pin_rx, !inverted));```. Note that it was changed recently and the codepath used by who added it (STM32F3xx) does not go through the STM32F1xx codepath so it was possible that it was inadvertently broken. Note that in the current code the RX pin is set to push-pull output (0x9) and not input (0x8) yet still manages to receive some data. I've gotten a bit off topic but just know that STLINK is the only way to update the FM30 with this PR.
* A bunch of extra GPIO need to be configured and I just dumped them into setup() which you may have better ideas for where to put this. Let me know if you'd like it refactored somehow before accepting.
* I renamed the GPIO `BUFFER_OE` to `GPIO_PIN_RC_INVERT`, since buffer_oe doesn't indicate what it is and buffer_oe_active doesn't describe what direction active should be (active for RX or active for TX?).
* The FM30 doesn't have a BUSY pin connected from the SX1280. Therefore I had to define-out all the busy-wait code. Busy is really only needed to guard against trying to do stuff during TX/RX/SBY transitions and ELRS doesn't try to do work immediately after these transitions (within 60uS) so it seems ok for the FM30 to go without this check.
* Adding a variant was needed due to the use of a 16MHz crystal which needed a different clock setup.
* There's only 100mW max power. 150mW is possible with the hardware, but I've limited the max to 100 since that extra power won't get you any more actual range.
* There is no external EEPROM so this uses the existing EEPROM emulation library. This library is total garbage. **Every single byte written** causes the entire 1KB page to be erased and rewritten. Not only that, it writes to the 127KBth flash page on devices that only have 64KB of flash so this is not guaranteed to work at all. I'll rewrite this if the PR is accepted.

FM30 mini RX support is next on my TODO list, so I can't test if this works, but I can tell you that an RF power meter shows power spikes at the right frequency and ELRS LUA reports 0:250 so everything seems to be in order.

cruwaller commented on src/platformio.ini:
```diff
@@ -58,8 +58,6 @@ build_flags =
 	-D PLATFORM_STM32=1
 	-O2
 src_filter = ${common_env_data.src_filter} -<ESP32*.*> -<ESP8266*.*> -<WS281B*.*>
-lib_deps =
```
revert these changes. you can just define empty `lib_deps =` if you want to remove this.

cruwaller commented on src/src/targets.h:
```diff
@@ -194,8 +194,8 @@ Credit to Jacob Walser (jaxxzer) for the pinout!!!
 
-#define BUFFER_OE               PA5  //CONFIRMED
-#define BUFFER_OE_ACTIVE HIGH
+#define GPIO_PIN_RC_INVERT      PA5  //CONFIRMED
```
this naming is not very suitable. That pin is also used to control buffer directions (originally).

CapnBry replied:
So what do you suggest then? `GPIO_PIN_RCSIGNAL_TX` is already used to name the TX GPIO, what's a name for a secondary pin that enabled TX? How about `GPIO_PIN_RC_TX_EN` and `GPIO_PIN_RC_TX_EN_INVERT`?

cruwaller replied:
myabe `GPIO_PIN_RCSIGNAL_RX_BUFFER` and `GPIO_PIN_RCSIGNAL_TX_BUFFER`.
Those follow the current RCSIGNAL_PIN naming :)
and `_INVERT` suffix...
@AlessandroAU any comment?

cruwaller commented on src/src/tx_main.cpp:
```diff
@@ -655,7 +656,22 @@ void setup()
+#endif
+
+#if defined(TARGET_TX_FM30)
+  pinMode(PB1, OUTPUT); // Green LED on "Red" LED (off)
```
Check the comments!

cruwaller commented on src/src/targets.h:
```diff
@@ -194,8 +194,8 @@ Credit to Jacob Walser (jaxxzer) for the pinout!!!
-#define BUFFER_OE               PA5  //CONFIRMED
-#define BUFFER_OE_ACTIVE HIGH
+#define GPIO_PIN_RC_INVERT      PA5  //CONFIRMED
+#define GPIO_PIN_RC_INVERT_TX   HIGH
```
Use 0 or 1 and later xor it.

CapnBry replied:
That is how it is in develop HEAD already, I did not write that I just renamed the define. I'll update it to use 0/1 pending naming of the define above.

cruwaller replied:
yeah, others need to be changed as well at some point :)

CapnBry commented:
I've talked with Alessandro on Discord development channel about the define name and he accepted GPIO_PIN_BUFFER_OE which is good for me too. I've updated the define as well as the INVERT to do an XOR with an 0/1 like it does everywhere else too.

This also adds BUSY delay to the hal. Once I got the RX side coded up I found that the TX had issues. I've measured the BUSY delay using an oscilloscope and added a deferred delay that gets realized in WaitNotBusy() and it should be completely optimized out for other targets that have a BUSY pin. I've run hour-long connection tests now with no LOS.

Let me know if there's anything else you'd like me to address @cruwaller and I appreciate your time in reviewing this (along with the smarter emulated EEPROM changes you've made).

cruwaller reviewed (changes requested):
Thanks! This looks good except the pin config in tx main.

CapnBry commented:
Haha yes it is a bit confusing the LED names eh? The LEDs in the module are bi-color LED with red and green in each physical LED so the LED we use as "Red" also has a green that needs to be kept off to keep it from being orange. Same thing with the "Green" LED that has an unused red component.

I've updated the code to use descriptive defines for the extra pins.

cruwaller reviewed (approved):
Thanks 👍

CapnBry commented:
I know you'd already approved this, but I thought it would be better to have some sort of functioning bootloader. This replaces the standard bootloader with a DFU version, so the user can just hold down the button, plug the USB in, and flash. The device shows up as "ExpressLRS DFU bootloader", I hope that's an acceptable use of the project name.

The code is from [stm32-dfu-bootloader](https://github.com/davidgfnet/stm32-dfu-bootloader), slightly modified to work with the FM30 hardware.

CapnBry commented:
Realized that the bluetooth was actually active-low so it was enabled (although idle, so <10mA of power but not doing anything). I've also switched the debug message UART to use the USB serial port via USB CDC / VCP and updated the DFU uploader build setup to reset after uploading.

cruwaller reviewed (approved):
Nice idea to use DFU :) 
great work!

AlessandroAU commented:
Thank you for all the work, I want to latency test my stock FM30 then I will flash it and confirm it's all working :)

CapnBry commented:
I know there are conflicts on this due to the devel branch code continuously changing, but just let me know when this is being considered for merge and I'll bring it up to date,

jurgelenas commented:
@CapnBry  I think you should also add your new target to [https://github.com/ExpressLRS/ExpressLRS/blob/47ec7b4347b02feda72ffe49f3be8cbad2ab5ffd/.github/workflows/build.yml](.github/workflows/build.yml).

CapnBry commented:
> @CapnBry I think you should also add your new target to [https://github.com/ExpressLRS/ExpressLRS/blob/47ec7b4347b02feda72ffe49f3be8cbad2ab5ffd/.github/workflows/build.yml](.github/workflows/build.yml).

Good call. Updated!
